src/xo-app/vm/tab-disks.js

```diff
@@ -1,146 +1,481 @@
-                              btnStyle='default'
-                              icon='vdi-remove'
-                              handler={deleteVdi}
-                              handlerParam={vdi}
-                            />
-                          </ButtonGroup>
-                        </span>
-                      }
-                    </td>
-                  </tr>
-                })}
-              </tbody>
-            </table>
-          </div>
+
+const writableSrPredicate = sr => sr.content_type !== 'iso'
+
+@propTypes({
+  onClose: propTypes.func,
+  vbds: propTypes.array.isRequired,
+  vdis: propTypes.object.isRequired,
+  vm: propTypes.object.isRequired
+})
+class NewDisk extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      sr: undefined
+    }
+  }
+
+  _createDisk = () => {
+    const { vm, vbds, vdis, onClose = noop } = this.props
+    const {name, size, bootable, readOnly} = this.refs
+    const { sr } = this.state
+    return createDisk(name.value, size.value, sr)
+      .then(diskId => {
+        const mode = readOnly.value ? 'RO' : 'RW'
+        let lastPos = 0
+        forEach(vbds, vbd => {
+          if (vdis[vbd.VDI]) {
+            lastPos = Math.max(lastPos, +vbd.position)
+          }
+        })
+        return attachDiskToVM(vm, diskId, {
+          bootable: bootable.value,
+          mode,
+          position: lastPos + 1
+        })
+          .then(onClose)
+      })
+  }
+
+  _selectSr = sr => this.setState({sr})
+
+  render () {
+    return <form id='newDiskForm'>
+      <div className='form-group'>
+        <SelectSr predicate={writableSrPredicate} onChange={this._selectSr} required />
+      </div>
+      <fieldset className='form-inline'>
+        <div className='form-group'>
+          <input type='text' ref='name' placeholder='Disk Name' className='form-control' required />
+        </div>
+        {' '}
+        <div className='form-group'>
+          <SizeInput ref='size' placeholder='Size' required />
+        </div>
+        {' '}
+        <div className='form-group'>
+          Bootable <Toggle ref='bootable' /> Readonly <Toggle ref='readOnly' />
+        </div>
+        <span className='pull-right'>
+          <ActionButton form='newDiskForm' icon='add' btnStyle='primary' handler={this._createDisk}>Create</ActionButton>
         </span>
-        : <h4 className='text-xs-center'>{_('vbdNoVbd')}</h4>
+      </fieldset>
+    </form>
+  }
+}
+
+@propTypes({
+  onClose: propTypes.func,
+  vbds: propTypes.array.isRequired,
+  vdis: propTypes.object.isRequired,
+  vm: propTypes.object.isRequired
+})
+class AttachDisk extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {}
+  }
+
+  _vdiPredicate = vdi => {
+    const { vm } = this.props
+    return vdi.$pool === vm.$pool
+  }
+
+  _srPredicate = sr => {
+    const { vm } = this.props
+    return sr.$pool === vm.$pool && sr.SR_type !== 'iso'
+  }
```

Comment:
Idem, it usually better to use a selector to recreate the predicate when the dependency changes.